File: sphone-api/Domain/Products/Product.cs
using SPhone.Domain.Common;

namespace SPhone.Domain.Products;

public sealed class Product : AggregateRoot<ProductId>
{
    public string Name { get; private set; }
    public string Description { get; private set; }
    public string Brand { get; private set; }
    public string ImageUrl { get; private set; }
    public Money Price { get; private set; }
    public ProductCategory Category { get; private set; }
    public bool IsAvailable { get; private set; }
    public DateTime CreatedAt { get; private set; }

    // EF Core constructor
    private Product() : base(ProductId.New())
    {
        Name = null!;
        Description = null!;
        Brand = null!;
        ImageUrl = null!;
        Price = null!;
    }

    private Product(
        ProductId id,
        string name,
        string description,
        string brand,
        string imageUrl,
        Money price,
        ProductCategory category) : base(id)
    {
        Name = name;
        Description = description;
        Brand = brand;
        ImageUrl = imageUrl;
        Price = price;
        Category = category;
        IsAvailable = true;
        CreatedAt = DateTime.UtcNow;
    }

    public static Product Create(
        string name,
        string description,
        string brand,
        string imageUrl,
        Money price,
        ProductCategory category)
    {
        if (string.IsNullOrWhiteSpace(name)) throw new DomainException("Product name is required.");
        if (string.IsNullOrWhiteSpace(brand)) throw new DomainException("Brand is required.");
        return new Product(ProductId.New(), name, description, brand, imageUrl, price, category);
    }

    public void SetAvailability(bool isAvailable) => IsAvailable = isAvailable;

    public void UpdatePrice(Money newPrice) => Price = newPrice;
}

File: sphone-api/src/domain/products/Product.ts
import { AggregateRoot } from '../shared/AggregateRoot'
import { ProductId } from './ProductId'
import { Money } from './Money'
import { ProductCategory } from './ProductCategory'

export class Product extends AggregateRoot {
  private constructor(
    private readonly _id: ProductId,
    private _name: string,
    private _description: string,
    private _brand: string,
    private _imageUrl: string,
    private _price: Money,
    private _category: ProductCategory,
    private _isAvailable: boolean,
    private readonly _createdAt: Date,
  ) {
    super()
  }

  get id(): ProductId { return this._id }
  get name(): string { return this._name }
  get description(): string { return this._description }
  get brand(): string { return this._brand }
  get imageUrl(): string { return this._imageUrl }
  get price(): Money { return this._price }
  get category(): ProductCategory { return this._category }
  get isAvailable(): boolean { return this._isAvailable }
  get createdAt(): Date { return this._createdAt }

  static reconstitute(params: {
    id: string
    name: string
    description: string
    brand: string
    imageUrl: string
    priceAmount: number
    priceCurrency: string
    category: number
    isAvailable: boolean
    createdAt: Date
  }): Product {
    return new Product(
      ProductId.create(params.id),
      params.name,
      params.description,
      params.brand,
      params.imageUrl,
      Money.create(params.priceAmount, params.priceCurrency),
      params.category as ProductCategory,
      params.isAvailable,
      params.createdAt,
    )
  }
}
